sda_download/sda_download.go

```diff
@@ -0,0 +1,259 @@
+`
+
+// ArgHelp is the suffix text that will be displayed after the argument list in
+// the module help
+var ArgHelp = `
+	[dataset]
+		The ID of the dataset that the file is part of.
+	[uri]
+		All flagless arguments will be used as sda-download uri.
+	[filename(s)]
+		The name of the file to download.`
+
+// Args is a flagset that needs to be exported so that it can be written to the
+// main program help
+var Args = flag.NewFlagSet("sda-download", flag.ExitOnError)
+
+var configPath = Args.String("config", "",
+	"S3 config file to use for downloading.")
+
+var datasetID = Args.String("dataset", "",
+	"Dataset ID for the file to download")
+
+var url = Args.String("url", "",
+	"The name of the file to download")
+
+var outDir = Args.String("outdir", "",
+	"Directory for downloaded files.")
+
+// necessary for mocking in testing
+var getResponseBody = getBody
+
+// File struct represents the file metadata
+type File struct {
+	FileID                    string `json:"fileId"`
+	DatasetID                 string `json:"datasetId"`
+	DisplayFileName           string `json:"displayFileName"`
+	FilePath                  string `json:"filePath"`
+	FileName                  string `json:"fileName"`
+	FileSize                  int    `json:"fileSize"`
+	DecryptedFileSize         int    `json:"decryptedFileSize"`
+	DecryptedFileChecksum     string `json:"decryptedFileChecksum"`
+	DecryptedFileChecksumType string `json:"decryptedFileChecksumType"`
+	FileStatus                string `json:"fileStatus"`
+	CreatedAt                 string `json:"createdAt"`
+	LastModified              string `json:"lastModified"`
+}
+
+// SdaDownload function downloads files from the SDA by using the
+// download's service APIs
+func SdaDownload(args []string) error {
+	var files []string
+	// Call ParseArgs to take care of all the flag parsing
+	err := helpers.ParseArgs(args, Args)
+	if err != nil {
+		return fmt.Errorf("failed parsing arguments, reason: %v", err)
+	}
+
+	if *datasetID == "" || *url == "" || *configPath == "" {
+		return fmt.Errorf("missing required arguments, dataset, config and url are required")
+	}
+
+	// Check that input file/folder list is not empty
+	if len(Args.Args()) == 0 {
+		return errors.New("no files to download")
+	}
+
+	files = append(files, Args.Args()...)
+
+	// Get the configuration file or the .sda-cli-session
+	config, err := helpers.GetAuth(*configPath)
+	if err != nil {
+		return err
+	}
+
+	// Check if the token has expired
+	err = helpers.CheckTokenExpiration(config.AccessToken)
+	if err != nil {
+		return err
+	}
+
+	// Loop through the files and download them
+	for _, filePath := range files {
+		downloadurl, err := downloadURL(*url, config.AccessToken, *datasetID, filePath)
+		if err != nil {
+			return err
+		}
+
+		filePathSplit := strings.Split(filePath, "/")
+		if strings.Contains(filePath, "elixir-europe.org") {
+			filePath = strings.Join(filePathSplit[1:], "/")
+		}
```

Comment:
This can't be hard coded

Reply:
In that part I am just checking if the userid is in the filepath in order not to remove and not making it part of the download path. Unless you mean that even that information is too much to exist there

Reply:
I mean the `elixir-europe.org` part. Because that will change depending on which instance you authenticate to.

Reply:
That will fail in FEGA where the user ID is of the format USER.NAME@HOST